New Monte Carlo Tree Search

urolyi opened the pull request.

urolyi commented:
Still need to debug a bit of the MCTS algorithm, but this should give some context on how I want to use the skyjo implementation. Hopefully, this will provide some more useful context on how the game will be "simmed" and what sort of ergonmics are useful.

I added some convenience functions to `skyjo.py` although I haven't thought super hard about making sure everything is still super performant. 

Still a draft, but feel free to take a look and make any suggestions/comments.

noahbkim commented on skyjo.py:
```diff
@@ -651,6 +734,42 @@ def greedy(skyjo: Skyjo) -> int:
+
+
+def apply_action(skyjo: Skyjo, action: SkyjoAction, rng: Random = random) -> Skyjo:
+    skyjo = randomize(skyjo, rng=rng)
```
you unfortunately have to write this out every time except for when you are replacing an already-flipped card, in which case there's no random even necessary

noahbkim commented on skyjo.py:
```diff
@@ -597,35 +644,71 @@ def actions(skyjo: Skyjo) -> np.ndarray[tuple[int], np.uint8]:
+    if MASK_FLIP <= action < MASK_FLIP + FINGER_COUNT:
+        return True
+    row, column = divmod(action - MASK_REPLACE, COLUMN_COUNT)
+    return table[0, row, column, FINGER_HIDDEN]
```
`bool()` this it's an int

noahbkim reviewed (commented):
I think it's probably simpler and marginally more efficient to just use the mask everywhere

noahbkim commented on skyjo.py:
```diff
@@ -77,6 +78,17 @@
+SkyjoAction: TypeAlias = int
+"""Integer representing an action in the Skyjo game."""
+
+StateValue: TypeAlias = np.ndarray[tuple[int], np.float32]
```
this might actually belong in MCTS or skynet

urolyi commented on skyjo.py:
```diff
@@ -312,12 +324,47 @@ def get_winner(skyjo: Skyjo) -> int:
+        (
+            skyjo[0].tobytes(),
+            skyjo[1].tobytes(),
+            skyjo[2].tobytes(),
```
It seems like hash(arr.data) directly doesn't work even after setting it as read-only.
```
ValueError: memoryview: hashing is restricted to formats 'B', 'b' or 'c'
```

Also, hash(arr.tostring()) yields a warning that it's deprecated and to use tobytes() instead. I ran a timeit experiment and they ran in identical times.

The other suggestions in the thread look wrong or not that useful a lot of discussion of using the shortened string rep which seems bad for our case.

I will note that this tobytes() will yield same hashes for arrays of different shape, but that shouldn't be a problem for us. I will add a comment though noting this.

urolyi commented:
- Reverted to using mask and then computing action list using np.argwhere
- Address other small comments
- **Fixed bug in end of round reveal that didn't rotate board to flip for other players (this wasn't caught in selfplay runs becuase the greedy strategy reveals all card before the end).**
- **Added end round countdown to skyjo state**
- Improved `apply_action` to simplify use and simplified `selfplay` using this
- Added functionality to reshuffle discard pile when deck is exhausted

urolyi commented on skyjo.py:
```diff
@@ -651,6 +819,88 @@ def greedy(skyjo: Skyjo) -> int:
+    return skyjo
+
+
+def apply_action(skyjo: Skyjo, action: SkyjoAction, rng: Random = random) -> Skyjo:
```
Simplifies the experience of callers to just mainly have to interact with this function rather than have to check any game conditions themselves.

urolyi commented on skyjo.py:
```diff
@@ -302,31 +355,46 @@ def get_is_visible(skyjo: Skyjo, player: int = 0) -> bool:
-                if table[player, row, column, card]:
-                    score += card - 2
-                    break
+    return (
```
Tested this and it's marginally faster likely because `.sum()` and `np.argwhere` are faster(?)

urolyi replied:
🙈

urolyi commented on skyjo.py:
```diff
@@ -302,31 +355,46 @@ def get_is_visible(skyjo: Skyjo, player: int = 0) -> bool:
 
-    return score
 
+def get_round_scores(
```
Added round ending doubling

noahbkim commented on mcts_new.py:
```diff
@@ -0,0 +1,280 @@
+
+# MARK: Enums
+class MCTSSelectionMethod(enum.Enum):
+    UCB = "UCB"  # AlphaZero UCB
```
I feel like subclassing might be more ergonomic/easier to separate into individual files/notebooks here

noahbkim commented on mcts_new.py:
```diff
@@ -0,0 +1,280 @@
+class DecisionStateNode:
+    state: sj.Skyjo
+    parent: AfterStateNode | None
+    prev_action: sj.SkyjoAction | None
```
this is of course an absurd and functionally meaningless nit, but if we plan on publishing this, I would prefer names to be unabbreviated; for example, `previous_action` here and `visit_count` below. I just think it reads cleaner idk

noahbkim commented on mcts_new.py:
```diff
@@ -0,0 +1,280 @@
+        self.model_output = model.predict(self.state)
+        self.model_output.policy_output = (
+            self.model_output.policy_output.mask_invalid_and_renormalize(
+                einops.rearrange(sj.actions(self.state), "a -> 1 a")
```
I know absolutely nothing about this package but am wary at the implication that it's reparsing this equation string on every invocation--is there a way to precompile this operation?

urolyi replied:
I will look into this some more, but I generally really like this package for the readability and intuitiveness once you learn the basics. I don't think it's that much slower than manually applying the same reshape and transpose operations, but will get back with some experiements to back that up.

noahbkim commented on mcts_new.py:
```diff
@@ -0,0 +1,280 @@
+        raise ValueError("Terminal nodes should not need to be expanded")
+
+
+MCTSNode = DecisionStateNode | AfterStateNode | TerminalStateNode
```
```suggestion
MCTSNode: TypeAlias = DecisionStateNode | AfterStateNode | TerminalStateNode
```

plus `from typing import TypeAlias`

plays better with mypy if we ever enable that

noahbkim commented on skyjo.py:
```diff
@@ -74,8 +75,28 @@
+"""
+
+SkyjoAction: TypeAlias = int
+"""Integer representing an action in the Skyjo game."""
```
Nothing to do with this particular PR, but consider doing this for `int -> Card` and adding documentation about how it's an index not a value (I wonder if `NewType` would be good here)

noahbkim commented on skyjo.py:
```diff
@@ -170,14 +192,37 @@ def _rotate_table(table: Table, players: int) -> None:
+                # Add cleared cards to discard pile
+                _push_top(game, j)
+                _push_top(game, j)
+                _push_top(game, j)
```
Oops

noahbkim commented on skyjo.py:
```diff
@@ -285,39 +355,95 @@ def get_is_visible(skyjo: Skyjo, player: int = 0) -> bool:
+    players = skyjo[3]
+    base_scores = np.array([get_score(skyjo, player=i) for i in range(players)])
+    round_ender_score = base_scores[round_ending_player]
+    base_scores[round_ending_player] = 1000  # larger than any other possible score
```
more foolproof to just use `inf`? feels like there should be a cooler way to do this :(

noahbkim commented on skyjo.py:
```diff
@@ -374,14 +500,20 @@ def randomize(skyjo: Skyjo, rng: Random = random) -> Skyjo:
+    # Deck is empty, reset with discarded cards
+    if not deck.any():
+        deck = game[GAME_DISCARDS : GAME_DISCARDS + CARD_SIZE]
+        game[GAME_DISCARDS : GAME_DISCARDS + CARD_SIZE] = 0
```
damn this ends up being super clean, nice

noahbkim commented on skyjo.py:
```diff
@@ -170,14 +192,37 @@ def _rotate_table(table: Table, players: int) -> None:
+    )
+
+
+def _clear_columns(table: Table, game: Game) -> int | None:
```
mega nit: order these same as `Skyjo` so `game, table`

noahbkim commented on skyjo.py:
```diff
@@ -613,21 +752,50 @@ def actions(skyjo: Skyjo) -> np.ndarray[tuple[int], np.uint8]:
             row, column = divmod(i, COLUMN_COUNT)
             if not table[0, row, column, FINGER_CLEARED]:
-                mask[MASK_REPLACE + i] = 1  # You can only replace
+                mask[MASK_REPLACE + i] = 1
```
not fuckin with comments huh

urolyi replied:
cursor doesn't fuck with comments ig

noahbkim commented on skyjo.py:
```diff
@@ -613,21 +752,50 @@ def actions(skyjo: Skyjo) -> np.ndarray[tuple[int], np.uint8]:
             row, column = divmod(i, COLUMN_COUNT)
             if table[0, row, column, FINGER_HIDDEN]:
-                mask[MASK_FLIP + i] = 1  # You both flip or replace
+                mask[MASK_FLIP + i] = 1
```
we can probably just inline these now

```suggestion
                mask[MASK_FLIP + I] = mask[MASK_REPLACE + i] = 1
```

noahbkim commented on skyjo.py:
```diff
@@ -613,21 +752,50 @@ def actions(skyjo: Skyjo) -> np.ndarray[tuple[int], np.uint8]:
     return mask
 
 
+def actions_list(skyjo: Skyjo) -> Iterable[SkyjoAction]:
```
```suggestion
def get_actions(skyjo: Skyjo) -> Iterable[SkyjoAction]:
```

noahbkim commented on skyjo.py:
```diff
@@ -613,21 +752,50 @@ def actions(skyjo: Skyjo) -> np.ndarray[tuple[int], np.uint8]:
+    # end of round reveals of facedown cards which will reveal all
+    if (
+        countdown is not None
+        and countdown == 1
```
wait, should this be 1 or 0? 0 feels more correct to me

noahbkim commented on skyjo.py:
```diff
@@ -651,6 +819,88 @@ def greedy(skyjo: Skyjo) -> int:
+                    skyjo = randomize(skyjo, rng=rng)
+                    skyjo = flip(skyjo, row, column, rotate=False, turn=False)
+                    assert validate(skyjo)
+        skyjo = _rotate_skyjo(skyjo)
```
hehe

noahbkim reviewed (approved):
All comments are fluff or nits, absolutely nothing of substance, feel free to to disregard and merge

urolyi commented on skyjo.py:
```diff
@@ -285,39 +355,95 @@ def get_is_visible(skyjo: Skyjo, player: int = 0) -> bool:
+def get_round_scores(
+    skyjo: Skyjo, round_ending_player: int = 0
+) -> np.ndarray[tuple[int], np.uint8]:
+    """Get the scores of all players for the current round."""
```
Btw I really like review comments like this as I am actively trying to get good "style". Feel free to be extra opinionated and nitty on stuff like this.
